pre_submit.csh

```diff
@@ -0,0 +1,208 @@
+#    assim only
+#    flexible numbers of cycles/job and resubmissions
+# and to check whether the case is ready for that submission:
+#    prevent running assimilations when there are too many cesm.log files in $rundir
+#    check the requested start date against the one in rpointer.atm_0001
+#    
+# Run in $CASEROOT
+
+if ($#argv == 0) then
+   echo "Usage: submit first_date last_date cycles_per_job queue"
+   echo "       first_date, last_date = YYYY-MM-DD-SSSSS"
+   echo "       first_date = last_date -> assimilation only."
+   echo "       cycles_per_job: The number of jobs will be calculated from the dates"
+   echo "                       Set to 1 for assimilation only jobs."
+   echo "       If 'queue' is omitted, the wall clock will be calculated, then exit"
+   exit
+endif
+
+set first_date     = $1
+set last_date      = $2
+set cycles_per_job = $3
+
+set all_inst = `./xmlquery NINST --value`
+set parts = `echo $all_inst[1] | sed -e "s#'# #g;s#:# #"`
+set num_inst = $parts[3]
+setenv case_run_dir `./xmlquery RUNDIR --value`
+set case_py_dir = '/glade/work/raeder/Models/cesm2_1_relsd_m5.6/cime/scripts/lib/CIME/case'
+
+# Check whether there are too many cesm.log files in rundir.
+set num_logs = `ls -1 ${case_run_dir}/cesm.log* | wc -l`
+
+if ($first_date == $last_date && $num_logs > 3 || \
+    $first_date != $last_date && $num_logs > 2 ) then
+   echo "ERROR: too many cesm.log files in $case_run_dir; remove the extraneous"
+   exit
+endif
+
+# Make sure that requested initial date is what CAM will use.
+if (-f ${case_run_dir}/rpointer.atm_0001) then
+   grep $first_date ${case_run_dir}/rpointer.atm_0001
+   if ($status != 0) then
+#     $first_date != $last_date && 
+      sed -e "s#NO-DATE-YET#$first_date#" stage_cesm_files.template >! stage_cesm_files
+      if ($status != 0) then
+         echo "ERROR: rpointer.atm_0001 has the wrong date, but creating stage_cesm_files failed"
+         exit
+      endif
+      echo "Running stage_cesm_files for $first_date"
+      chmod 755 stage_cesm_files
+      ./stage_cesm_files
+   endif
+endif
+
+# Translate dates into cycles.
+set start = `echo $first_date | sed -e "s#-# #g"`
+set end   = `echo $last_date  | sed -e "s#-# #g"`
+
+set days_in_mo = (31 28 31 30 31 30 31 31 30 31 30 31)
+if (($start[1] % 4) == 0 ) set days_in_mo[2] = 29
+
+@ years  = $end[1] - $start[1]
+@ months = $end[2] - $start[2]
+@ days   = $end[3] - $start[3]
+@ secs   = $end[4] - $start[4]
+if ($secs < 0 ) then
+   @ secs = $secs + 86400
+   @ days--
+endif
+if ($days < 0 ) then
+   @ days = $days + $days_in_mo[$start[2]]
+   @ months--
+endif
+if ($months < 0 ) then
+   @ months = $months + 12
+   @ years--
+endif
+
+if (($end[4] < $start[4] && $end[4] != '00000') || \
+    ($end[4] > $start[4] && $end[3] != $start[3]) ) then
+   echo "ERROR: submit requires an integral number of days "
+   echo "       or starting and ending on the same day"
+   echo "       or ending on 00000 of a future day"
+   exit
+endif
```

Comment:
UGH.   Can we use advance_time to do this? It handles leap years, seconds, etc. perfectly.

Reply:
with this code if you do runs for year 2000, it fails.   

advance_time is a good way to handle +/- time in scripts.  in addition to all the other options, can output CESM-style dates if that's useful anywhere.

Reply:
Advance time sounds like a good choice.
I only need it every couple years, so it's not the first thing that occurs to me.
Does this need to be changed and tested now?

I believe that it is correct for 2000.  Years divisible by 400 are leap years,
even though they are also divisible by 100, which would not be.

Reply:
my opinion is this should *not* be changed until after the reanalysis is running again.  but in the subsequent code reviews and cleanups, we should see about changing any and all script-based time computations to use advance_time.

(everyone's right that 2000 was indeed a leap year - i got the modulo 400 thing backwards in my head.)